codegenerator/cli/templates/static/codegen/src/eventFetching/FetchState.res

```diff
@@ -875,22 +875,19 @@ fetching registration
 let isReadyForNextQuery = (self: t, ~maxQueueSize) =>
   self.baseRegister.fetchedEventQueue->Array.length < maxQueueSize
 
-let rec getAllAddressesForContract = (
-  ~addresses=Set.String.empty,
+let rec checkBaseRegisterContainsRegisteredContract = (
+  register: register,
   ~contractName,
-  self: register,
+  ~contractAddress,
 ) => {
-  let addresses =
-    self.contractAddressMapping
-    ->ContractAddressingMap.getAddresses(contractName)
-    ->Option.mapWithDefault(addresses, newAddresses => {
-      addresses->Set.String.union(newAddresses)
-    })
-
-  switch self.registerType {
-  | RootRegister(_) => addresses
-  | DynamicContractRegister({nextRegister}) =>
-    nextRegister->getAllAddressesForContract(~addresses, ~contractName)
+  switch register.contractAddressMapping->ContractAddressingMap.getAddresses(contractName) {
+  | Some(addresses) if addresses->Belt.Set.String.has(contractAddress->Address.toString) => true
+  | _ =>
+    switch register.registerType {
+    | RootRegister(_) => false
+    | DynamicContractRegister({nextRegister}) =>
+      nextRegister->checkBaseRegisterContainsRegisteredContract(~contractName, ~contractAddress)
+    }
```

Comment:
Improved this function to early return rather than always iterate over all registers